tools/cassandra_delete_range/main.go

```diff
@@ -0,0 +1,502 @@
+
+	// ledger_transactions queries
+	log.Println("Generating delete queries for ledger_transactions table")
+	info, traversedRowsCount = prepareDeleteQueries(cluster,
+		fromLedgerIdx, toLedgerIdx,
+		"SELECT hash, ledger_sequence FROM ledger_transactions WHERE token(ledger_sequence) >= %s AND token(ledger_sequence) <= %s",
+		"DELETE FROM ledger_transactions WHERE hash = ? AND ledger_sequence = ?")
+	log.Printf("Total delete queries: %d\n", len(info.Data))
+	log.Printf("Total traversed rows: %d\n\n", traversedRowsCount)
+	performDeleteQueries(cluster, &info, columnSettings{UseBlob: true, UseSeq: true})
+
+	// ledgers queries
+	log.Println("Generating delete queries for ledgers table")
+	info, traversedRowsCount = prepareDeleteQueries(cluster,
+		fromLedgerIdx, toLedgerIdx,
+		"SELECT header, sequence FROM ledgers WHERE token(sequence) >= %s AND token(sequence) <= %s",
+		"DELETE FROM ledgers WHERE sequence = ?")
+	log.Printf("Total delete queries: %d\n", len(info.Data))
+	log.Printf("Total traversed rows: %d\n\n", traversedRowsCount)
+	performDeleteQueries(cluster, &info, columnSettings{UseBlob: false, UseSeq: true})
+
+	// TODO: tbd what to do with account_tx as it got tuple for seq_idx
+	// TODO: also, whether we need to take care of nft tables and other stuff like that
+
+	if err := updateLedgerRange(cluster, fromLedgerIdx-1); err != nil {
+		log.Fatalf("ERROR failed updating ledger range: %s\n", err)
+	}
+
+	return nil
+}
+
+func prepareDeleteQueries(cluster *gocql.ClusterConfig, fromLedgerIdx uint64, toLedgerIdx uint64, queryTemplate string, deleteQueryTemplate string) (deleteInfo, uint64) {
+	rangesChannel := make(chan *tokenRange, len(ranges))
+	for i := range ranges {
+		rangesChannel <- ranges[i]
+	}
+
+	close(rangesChannel)
+
+	outChannel := make(chan deleteParams)
+	var info = deleteInfo{Query: deleteQueryTemplate}
+
+	go func() {
+		for params := range outChannel {
+			info.Data = append(info.Data, params)
+		}
+	}()
+
+	var wg sync.WaitGroup
+	var sessionCreationWaitGroup sync.WaitGroup
+	var totalRows uint64
+
+	wg.Add(numberOfParallelClientThreads)
+	sessionCreationWaitGroup.Add(numberOfParallelClientThreads)
+
+	for i := 0; i < numberOfParallelClientThreads; i++ {
+		go func() {
+			defer wg.Done()
+
+			var session *gocql.Session
+			var err error
+			if session, err = cluster.CreateSession(); err == nil {
+				defer session.Close()
+
+				sessionCreationWaitGroup.Done()
+				sessionCreationWaitGroup.Wait()
+				preparedQueryString := fmt.Sprintf(queryTemplate, "?", "?")
+				preparedQuery := session.Query(preparedQueryString)
+
+				for r := range rangesChannel {
+					preparedQuery.Bind(r.StartRange, r.EndRange)
+					iter := preparedQuery.Iter()
+
+					var key []byte
+					var seq uint64
+					var rowsRetrieved uint64
+
+					for iter.Scan(&key, &seq) {
+						rowsRetrieved++
+
+						// only grab the rows that are in the correct range of sequence numbers
+						if fromLedgerIdx <= seq && seq <= toLedgerIdx {
```

Comment:
We actually can not guarantee the table does not contain the toLedgerIdx+1 . If the Clio stops in the middle of writing , the range table is not updated, but other tables might be wroten.

Reply:
Yea, there is little we can do there i think, there is no transactions or anything like that in cassandra.. any suggestions?

Reply:
I misread your comment at first. I think we can just take everything that is newer than `fromLedgerIndex`. 
Originally i thought i tcould be useful to remove any range but i think that may only make more problems 👍